File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n=nums.size();
        priority_queue<pair<int,int>>pq;
        int i=0;
        for(;i<k;i++){
            pq.push({nums[i],i});
        }
        vector<int>ans;
        ans.push_back(pq.top().first);
        while(i<n){
            while(!pq.empty() && pq.top().second<=i-k){
                pq.pop();
            }
            pq.push({nums[i],i});
            auto top=pq.top();
            int val=top.first;
            ans.push_back(val);
            i++;
        }
        return ans;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {  
        PriorityQueue<int[]>pq = new PriorityQueue<>((a,b)->{
            if(a[0]==b[0]) return a[1] - b[1]; // minheap;
            return b[0] - a[0];//maxheap a after b -> a is greater;
        });
        int n = nums.length;
        int []ans=new int[n-k+1];
        int index=0;
        int i;
        for(i=0;i<k;i++){
            pq.offer(new int[]{nums[i],i});
        }
        ans[index++] = pq.peek()[0];
        while(i<n){
            pq.offer(new int[]{nums[i],i});
            while(!pq.isEmpty() && pq.peek()[1] <= i-k){
                pq.poll();
            }
            ans[index++] = pq.peek()[0];
            i++;
    
    }
    return ans;
}
};